### PR TITLE
TRUNK-4769 : Add updateSearchIndexAsync method to Context

### PR DESCRIPTION
**Pull request by aj-jaswanth**
Added a new method to update the search index asynchronously.

### CODE REVIEW

**Comment by dkayiwa** [comment]
@aj-jaswanth this has merge conflicts. You need to first pull the latest changes from upstream.

**Comment by aj-jaswanth** [comment]
Currently I have this commit in a branch. Shall I merge this with master and then create a pull request from master?

**Comment by dkayiwa** [comment]
Pull the latest changes from master into that branch.

On Wed, Oct 28, 2015 at 3:12 PM, Venkata Jaswanth <notifications@github.com>
wrote:

> Currently I have this commit in a branch. Shall I merge this with master
> and then create a pull request from master?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openmrs/openmrs-core/pull/1605#issuecomment-151825885>
> .
>



-- 
If we keep uppermost in our minds the unkind and unjust acts of others, we
shall find it impossible to love them as Christ has loved us; but if our
thoughts dwell upon the wondrous love and pity of Christ for us, the same
spirit will flow out to others.

**Comment by aj-jaswanth** [comment]
@dkayiwa Fixed merge conflict.

**Comment by rkorytkowski** [comment]
What is the use case for this method?

**Comment by aj-jaswanth** [comment]
Without asynchronous search index update, it takes significant time to startup openmrs in development or in deployment if the system has a lot of concepts.

**Comment by lluismf** [comment]
Any long process is a potential candidate for asynchronous execution. But there is already a task executor available right?

**Comment by aj-jaswanth** [comment]
Yeah. But that will create a dependency on Spring. Moreover, MassIndexer has the async method. So, it would be good to provide a method, that uses MassIndexer's method behind the scenes.

File: api/src/main/java/org/openmrs/api/context/Context.java [hunk 1]
```diff
@@ -1272,6 +1273,21 @@ public static Properties getConfigProperties() {
...
 	}
+
+	/**
+	 * Updates the search index. It is an asynchronous operation.
+	 * <p>
+	 * There is no need to call this method in normal usage since the index is automatically updated
+	 * whenever DB transactions are committed.
+	 * <p>
+	 * The method is designated to be used in tests, which rollback transactions. Note that if the
+	 * transaction is rolled back, changes to the index will not be reverted.
```
Comment (rkorytkowski) [inline api/src/main/java/org/openmrs/api/context/Context.java]: Could you please fix the javadoc? It is not intended for tests rather for use by modules to update the index in background after any database changes done omitting OpenMRS API and Hibernate.

**Comment by rkorytkowski** [comment]
@aj-jaswanth, ok, let's add that, but please fix the javadoc.

**Comment by aj-jaswanth** [comment]
@rkorytkowski Fixed the javadoc. But, I have two commits now. Should I squash them and force commit?

**Comment by dkayiwa** [comment]
@aj-jaswanth yes go ahead and squash them into one.

**Comment by aj-jaswanth** [comment]
@dkayiwa  Squashed commits into one.

**Comment by aj-jaswanth** [comment]
@dkayiwa Well, I made few changes to the approach. Committed them now. As the final result (side effect) of the updateSearchIndexAsync method call is out of its scope, moreover, the method will be called as an API rather than, being called as a side effect on UI, therefore it is nice to leave the clean up to the callee.

**Comment by aj-jaswanth** [comment]
@dkayiwa Could you please back port the commit to 1.12 branch also? Thank you.

**Comment by dkayiwa** [comment]
@aj-jaswanth i tried to automatically cherry-pick it into 1.12 but got merge conflicts. So do you mind just creating a pull request on the 1.12 branch?

**Pull request merged**